metadata-ingestion/src/datahub/ingestion/source/sql/clickhouse.py

```diff
@@ -118,30 +124,92 @@ def __post_init__(self):
 
 
 class ClickHouseConfig(
-    BasicSQLAlchemyConfig, BaseTimeWindowConfig, DatasetLineageProviderConfigBase
+    TwoTierSQLAlchemyConfig, BaseTimeWindowConfig, DatasetLineageProviderConfigBase
 ):
     # defaults
     host_port = Field(default="localhost:8123", description="ClickHouse host URL.")
     scheme = Field(default="clickhouse", description="", hidden_from_docs=True)
     password: pydantic.SecretStr = Field(
         default=pydantic.SecretStr(""), description="password"
     )
-
     secure: Optional[bool] = Field(default=None, description="")
     protocol: Optional[str] = Field(default=None, description="")
+    _deprecate_secure = pydantic_field_deprecated("secure")
+    _deprecate_protocol = pydantic_field_deprecated("protocol")
 
+    uri_opts: Dict[str, str] = Field(
+        default={},
+        description="The part of the URI and it's used to provide additional configuration options or parameters for the database connection.",
+    )
     include_table_lineage: Optional[bool] = Field(
         default=True, description="Whether table lineage should be ingested."
     )
     include_materialized_views: Optional[bool] = Field(default=True, description="")
 
-    def get_sql_alchemy_url(self, database=None):
-        uri_opts = None
-        if self.scheme == "clickhouse+native" and self.secure:
-            uri_opts = {"secure": "true"}
-        elif self.scheme != "clickhouse+native" and self.protocol:
-            uri_opts = {"protocol": self.protocol}
-        return super().get_sql_alchemy_url(uri_opts=uri_opts)
+    def get_sql_alchemy_url(self, current_db=None):
+        url = make_url(
+            super().get_sql_alchemy_url(uri_opts=self.uri_opts, current_db=current_db)
+        )
+        if url.drivername == "clickhouse+native" and url.query.get("protocol"):
+            logger.debug(f"driver = {url.drivername}, query = {url.query}")
+            raise Exception(
+                "You cannot use a schema clickhouse+native and clickhouse+http at the same time"
+            )
+
+        self.scheme = url.drivername
+        self.username = url.username
+        self.password = (
+            pydantic.SecretStr(str(url.password))
+            if url.password
+            else pydantic.SecretStr("")
+        )
+        if url.host and url.port:
+            self.host_port = url.host + ":" + str(url.port)
+        elif url.host:
+            self.host_port = url.host
+        # untill released https://github.com/python/mypy/pull/15174
+        self.uri_opts = {str(k): str(v) for (k, v) in url.query.items()}
+
+        url = make_url(
+            make_sqlalchemy_uri(
+                self.scheme,
+                self.username,
+                self.password.get_secret_value() if self.password else None,
+                self.host_port,
+                current_db if current_db else self.database,
+                uri_opts=self.uri_opts,
+            )
+        )
+
+        return str(url)
```

Comment:
Not clear to me why we need these all lines. Also, I believe we can do all this without importing make_url. Please LMK if I'm missing anything.
Shouldn't there be just 2 things here
1. validation of uri_opts against scheme
2. return  results of super().get_sql_alchemy_url(uri_opts=self.uri_opts, current_db=current_db)